corehq/apps/app_manager/util.py

```diff
@@ -507,84 +508,102 @@ def get_sort_and_sort_only_columns(detail, sort_elements):
     return sort_only_elements, sort_columns
 
 
-def get_form_data(domain, app, include_shadow_forms=True):
-    from corehq.apps.reports.formdetails.readable import FormQuestionResponse
-    from corehq.apps.app_manager.models import ShadowForm
-
-    case_meta = app.get_case_metadata()
-    modules = []
-    errors = []
-    for module in app.get_modules():
-        forms = []
-        module_meta = {
+class FormDataGenerator(object):
+    def __init__(self, domain, app, include_shadow_forms=True):
+        self.domain = domain
+        self.app = app
+        self.include_shadow_forms = include_shadow_forms
+
+        self.errors = []
+
+        self._seen_save_to_case = defaultdict(list)
+        self._case_meta = self.app.get_case_metadata()
+
+    def get_app_data(self):
+        return [self._compile_module(module) for module in self.app.get_modules()], self.errors
+
+    def _compile_module(self, module):
+        return {
             'id': module.unique_id,
             'name': module.name,
             'short_comment': module.short_comment,
             'module_type': module.module_type,
             'is_surveys': module.is_surveys,
             'module_filter': module.module_filter,
+            'forms': [self._compile_form(form) for form in self._get_pertinent_forms(module)],
```

Comment:
I especially like this, where you split up getting the forms and interpreting them